Sample/EventsVersioning/EventsVersioning.Tests/ExplicitSerialization/ShoppingCart.cs

```diff
@@ -0,0 +1,413 @@
+using System.Text.Json;
+using System.Text.Json.Nodes;
+
+namespace EventsVersioning.Tests.ExplicitSerialization;
+
+using static ShoppingCartEvent;
+
+public abstract class StronglyTypedValue<T>: IEquatable<StronglyTypedValue<T>> where T : notnull
+{
+    public T Value { get; }
+
+    protected StronglyTypedValue(T value) => Value = value;
+
+    public override string ToString() => Value.ToString()!;
```

Comment:
that renders Guids with dashes etc, which can end up very messy - might be worth considering how to model where you want e.g. cart id not to have embedded dashes

Reply:
For me the problem is that leaning on the default ToString of a type when you're trying to represent something as typesafe is a hole - I'd tend to either have a base StringId class with explicit translations, or leave abstract methods etc. (see the other types I linked for ways to remove that ambiguity/implicitness).
You can still do some `CreateDefault` helper that plugs in the ToString explicitly afterwards

I guess if you can take Guids out of the picture entirely, then the issue of how to render them is off the table too...